yarn-project/aztec-nr/aztec/src/account.nr

```diff
@@ -0,0 +1,79 @@
+use crate::entrypoint::EntrypointPayload;
+use crate::context::{PrivateContext, PublicContext, Context};
+use crate::oracle::compute_selector::compute_selector;
+use crate::state_vars::{map::Map, public_state::PublicState};
+use crate::types::type_serialisation::field_serialisation::{FieldSerialisationMethods,FIELD_SERIALISED_LEN};
+use crate::auth::IS_VALID_SELECTOR;
+
+struct AccountActions {
+  context: Context,
+  is_valid_impl: fn(&mut PrivateContext, Field) -> bool,
+  approved_action: Map<PublicState<Field, FIELD_SERIALISED_LEN>>,
+}
+
+impl AccountActions {
+  fn init(context: Context, approved_action_storage_slot: Field, is_valid_impl: fn(&mut PrivateContext, Field) -> bool) -> Self {
+    AccountActions {
+      context: context,
+      is_valid_impl: is_valid_impl,
+      approved_action: Map::new(
+        context,
+        approved_action_storage_slot,
+        |context, slot| {
+          PublicState::new(context, slot, FieldSerialisationMethods)
+        },
+      ),
+    }
+  }
+
+  fn private(context: &mut PrivateContext, approved_action_storage_slot: Field, is_valid_impl: fn(&mut PrivateContext, Field) -> bool) -> Self {
+    AccountActions::init(Context::private(context), approved_action_storage_slot, is_valid_impl)
+  }
+
+  fn public(context: &mut PublicContext, approved_action_storage_slot: Field, is_valid_impl: fn(&mut PrivateContext, Field) -> bool) -> Self {
+    AccountActions::init(Context::public(context), approved_action_storage_slot, is_valid_impl)
+  }
+  
+  // docs:start:entrypoint
+  fn entrypoint(self, payload: EntrypointPayload) {
+    let message_hash = payload.hash();
+    let valid_fn = self.is_valid_impl;
+    let private_context = self.context.private.unwrap();
+    assert(valid_fn(private_context, message_hash));
+    payload.execute_calls(private_context);
+  }
+  // docs:end:entrypoint
+
+  fn is_valid(self, message_hash: Field) -> Field {
+    let valid_fn = self.is_valid_impl;
+    if (valid_fn(self.context.private.unwrap(), message_hash)) {
+        IS_VALID_SELECTOR
+    } else {
+        0
+    }
+  }
+
+  fn is_valid_public(self, message_hash: Field) -> Field {
+    let value = self.approved_action.at(message_hash).read();
+    if (value == 1){
+        IS_VALID_SELECTOR
+    } else {
+        0
+    }
+  }
+
+  // TODO: See if we can remove this function. Maybe we can have the entrypoint call directly into internal_set_is_valid_storage..?
+  fn set_is_valid_storage(self, message_hash: Field, value: Field) {
```

Comment:
> nothing -> true -> false

I guess it'll depend on whether we differentiate empty from zero in public state? Are we making that distinction now?

Reply:
initially 0, but I think it might get a little weird because of the larger storage tree that we are using 🤔 essentially inserting into it, and a removal would normally be to set the value to zero, but that value is still in the tree and it is different from nothing. I'm not fully sure at what point in public execution the database will actually be updated, @iAmMichaelConnor or @dbanks12 might know a bit here? Ideally the squashing should mean that there is nothing in the tx but I'm not completely sure when the first change is insertion.